stablehlo/reference/ProcessGrid.cpp

```diff
@@ -142,45 +160,61 @@ ProcessGroups ProcessGrid::flattenedIds(
   return processGroups;
 }
 
-std::mutex &ProcessGrid::getRendezvousLock(ProcessGroup processGroup,
-                                           ChannelId channelId) {
-  std::lock_guard<std::mutex> lock(rendezvousLock_);
-  std::pair<ProcessGroup, ChannelId> channelKey(processGroup, channelId);
-  return channelLocks_[channelKey];
-}
+void ProcessGrid::outfeed(ArrayRef<Tensor> inputs) { outfeed_.push(inputs); }
 
-void ProcessGrid::outfeed(ArrayRef<Tensor> inputs) {
-  std::lock_guard<std::mutex> lock(outfeedLock_);
-  outfeed_.emplace(inputs);
-}
-
-RendezvousResult ProcessGrid::rendezvous(ProcessGroup processGroup,
-                                         ChannelId channelId,
-                                         ProcessId processId,
-                                         const Tensor &operand) {
+std::shared_ptr<RendezvousResult> ProcessGrid::rendezvous(
```

Comment:
Here was my attempt to rewrite this in a more understandable way. There are two assumptions:
1. wait/notify respects the order of threads arriving, using a queue - I think this is the case.
2. state.values is not written to again until all processes have returned from this thread (can be tweaked to account for this).

```c++
std::shared_ptr<RendezvousResult> ProcessGrid::rendezvous(
    ProcessGroup processGroup, ChannelId channelId, ProcessId processId,
    const Tensor &operand) {
  std::pair<ProcessGroup, ChannelId> channelKey(processGroup, channelId);
  // Process wait/notify logic below doesn't work for single process.
  if (processGroup.size() == 1)
    return std::make_shared<RendezvousResult>(
        RendezvousResult({std::pair{processId, operand}}));

  auto &state = channels_[channelKey];

  std::unique_lock<std::mutex> lock(state.mutex);
  state.values[processId] = operand;

  if (state.values.size() == processGroup.size()) {
    // If values are full, that means all other processes are currently waiting.
    // The last process to contribute moves the values into the result
    // then waits for each process to return a copy of the result before
    // cleaning up the state variable for future computations in this process
    // grid.
    state.result = std::make_shared<RendezvousResult>(state.values);
    channelConditions_[channelKey].notify_one();
    if (channelConditions_[channelKey].wait_for(
            lock, std::chrono::seconds(3)) == std::cv_status::timeout)
      llvm::report_fatal_error(
          "rendezvous timed out: not all process has received the results yet");

    // Only woken up once all other processes in group have returned.
    // Clean up the shared state.
    state.values.clear();
    return std::move(state.result);
  }

  // The remaining processes wait for the last process to contribute to move
  // the values into the shared result.
  if (channelConditions_[channelKey].wait_for(lock, std::chrono::seconds(3)) ==
      std::cv_status::timeout)
    llvm::report_fatal_error(
        "rendezvous timed out: not all processes have contributed yet");

  if (state.result == nullptr)
    llvm::report_fatal_error("rendezvous expected result to be set");

  // Copy result from state before notifying.
  auto result = state.result;
  channelConditions_[channelKey].notify_one();
  return result;
}
```

Reply:
Updated the commit with some minor changes. PTAL